instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/http/HttpServiceEventsObserver.kt

```diff
@@ -0,0 +1,120 @@
+package com.mapbox.navigation.instrumentation_tests.utils.http
+
+import com.mapbox.common.DownloadOptions
+import com.mapbox.common.HttpRequest
+import com.mapbox.common.HttpResponse
+import com.mapbox.common.HttpServiceFactory
+import com.mapbox.common.HttpServiceInterceptorInterface
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.mapNotNull
+import java.net.URL
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.locks.Lock
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.withLock
+
+sealed class HttpServiceEvent {
+
+    abstract val request: HttpRequest
+
+    val url: URL
+        get() = URL(request.url)
+
+    data class Request(override val request: HttpRequest) : HttpServiceEvent()
+
+    data class Download(
+        val download: DownloadOptions,
+        override val request: HttpRequest = download.request
+    ) : HttpServiceEvent()
+
+    data class Response(
+        val response: HttpResponse,
+        override val request: HttpRequest = response.request
+    ) : HttpServiceEvent()
+}
+
+class HttpServiceEventsObserver : HttpServiceInterceptorInterface {
+
+    private val lock: Lock = ReentrantLock()
+    private val callbacks = CopyOnWriteArrayList<OnHttpEventCallback>()
+    private val _latestEvents = mutableListOf<HttpServiceEvent>()
+
+    val latestEvents: List<HttpServiceEvent>
+        // Return a copy so that returned reference points to a list which won't be modified later.
+        get() = _latestEvents.toList()
+
+    val onRequestEvents: List<HttpServiceEvent.Request>
+        get() = latestEvents.filterIsInstance(HttpServiceEvent.Request::class.java)
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    fun eventsFlow(): Flow<HttpServiceEvent> {
+        return callbackFlow {
+            val callback = object : OnHttpEventCallback {
+                override fun onHttpEvent(event: HttpServiceEvent) {
+                    trySend(event)
+                }
+            }
+
+            lock.withLock {
+                latestEvents.forEach { trySend(it) }
+                addOnHttpEventCallback(callback)
+            }
```

Comment:
It's not recommended to mix locks from java with coroutine as it's too easy to make a mistake: any `suspend` function could change thread so that different thread in the end will try to release lock.

Maybe it's better to use mutex from coroutines library? It has `withLock` extension function too. 

Another option is to replace flow by a shared flow with unlimited(if it's possible) replay cache: https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-shared-flow/

Reply:
I used Java Lock here because one of usages is inside non-suspend function. But `MutableSharedFlow` perfectly fits here, thank you!